### PR TITLE
Feature 1 create branch stub

### PR DESCRIPTION
**Pull request by kuzmaka**

### CODE REVIEW

File: Lakat/extension.json [hunk 1]
```diff
@@ -33,6 +33,12 @@
 		"BeforePageDisplay": "LakatHooks",
 		"SkinTemplateNavigation::Universal": "LakatHooks"
 	},
+	"SpecialPages": {
+		"CreateBranch": "MediaWiki\\Extension\\Lakat\\SpecialCreateBranch"
+	},
+	"ExtensionMessagesFiles": {
+		"LakatAlias": "i18n/Lakat.alias.php"
+	},
```
Comment (leomarlo) [inline Lakat/extension.json:41]: What does this do?

File: Lakat/i18n/Lakat.alias.php [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+$specialPageAliases = [];
+
+$specialPageAliases['en'] = [
+	'CreateBranch' => ['Create Branch'],
+];
```
Comment (leomarlo) [inline Lakat/i18n/Lakat.alias.php:7]: What is the idea behind this file? It's for language compatibility, right?

File: Lakat/src/Hooks.php [hunk 3]
```diff
@@ -19,6 +19,7 @@
 
 namespace MediaWiki\Extension\Lakat;
 
+use MediaWiki\Title\Title;
```
Comment (leomarlo) [inline Lakat/src/Hooks.php:22]: General question about php. This brings is some namespace stuff that brings the Title object into scope, right? How do i read this path? Is Title a submodule of MediaWiki and Title an Object in the module Title?

File: Lakat/src/SpecialCreateBranch.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
+<?php
```
Comment (leomarlo) [inline Lakat/src/SpecialCreateBranch.php:1]: Thanks. So what is the effect of defining a SpecialCreateBranch extension of FormSpecialPage? FormSpecialPage is I guess one of the many classes that is provided by MediaWiki to help build extensions?

**Review by leomarlo — APPROVED** [review]

File: Lakat/extension.json (hunk 1 above)
Comment (kuzmaka) [reply Lakat/extension.json:41]: This adds reference to the translations and alternative spellings of the title of the special page. If it is not provided then you get warnings on top of screen on some pages, e.g. on the page /wiki/Special:SpecialPages where all special pages are listed. Warnings say something like "You should provide alias for special page title". I've tried to use simple localization file for this, but warning didn't disappear without this additional file.

File: Lakat/i18n/Lakat.alias.php (hunk 2 above)
Comment (kuzmaka) [reply Lakat/i18n/Lakat.alias.php:7]: See comment above for explainations why this is necessary.

File: Lakat/src/Hooks.php (hunk 3 above)
Comment (kuzmaka) [reply Lakat/src/Hooks.php:22]: > This brings is some namespace stuff that brings the Title object into scope, right?

Right. This imports class `\MediaWiki\Title\Title` (fully qualified name of the class with full namespace prefix) into Hooks.php and gives it a short alias `Title` that can be used in code. Otherwise, without `use` statement you have to reference fully qualified class name in code.

> How do i read this path?

TLDR: classes from `\MediaWiki\` namespace can be found in `/includes/` dir. Root path of namespace is not necessarely the same as path in filesystem.

This thing is closely related to autoloading of classes. If autoloading is not used, then you have to directly include one php file into another. PHP uses autoloader function to calculate class file path based on it's namespace. PHP has default [autoloader](https://www.php.net/manual/ru/function.spl-autoload.php) that searches for classes in current dir and configured library include paths. Mediawiki, like many other projects, replaces PHP default autoloader with [its own](https://www.mediawiki.org/wiki/Manual:AutoLoader.php). In that file you can see mapping of namespaces to file paths for MediaWiki files:
```
'MediaWiki\\' => __DIR__ . '/',
```
where __DIR__ is `/includes/` dir in mediawiki installation.
You can also find mapping for each class individually in automatically generated autoload.php file:
```
'MediaWiki\\Title\\Title' => __DIR__ . '/includes/title/Title.php',
```

For external libraries that are installed with composer, composer creates it's own autoloader, all those files can be found in `/vendor/` dir in relevant library subdir.

> Is Title a submodule of MediaWiki and Title an Object in the module Title?

Kind of, but I wouldn't call it modules because PHP doesn't have real modules contrary to Python. In PHP world they're namespaces and classes in those namespaces. Namespaces are used to arrange classes just like directory hierarchy is used to arrange files.

You correctly assumed that's the common agreement is to use the name of the project as the first part of the namespace. Following parts are usually subdirectories in that project and reflect some logical arrangement of classes.

File: Lakat/src/SpecialCreateBranch.php (hunk 4 above)
Comment (kuzmaka) [reply Lakat/src/SpecialCreateBranch.php:1]: In mediawiki terms, page that is created programmatically is named [special page](https://www.mediawiki.org/wiki/Manual:Special_pages). When creating special page, the class should be inherited from SpecialPage class or one of its subclasses. I chose FormSpecialPage because we needed from and that class has some work already done. Many other special pages in mediawiki inherit FormSpecialPage, e.g. SpecialPasswordReset, I look at them as inspiration.

File: Lakat/extension.json (hunk 1 above)
Comment (leomarlo) [reply Lakat/extension.json:41]: Perfect. Thanks :)

File: Lakat/i18n/Lakat.alias.php (hunk 2 above)
Comment (leomarlo) [reply Lakat/i18n/Lakat.alias.php:7]: Thanks

File: Lakat/src/Hooks.php (hunk 3 above)
Comment (leomarlo) [reply Lakat/src/Hooks.php:22]: Fantastic, crystal clear!

File: Lakat/src/SpecialCreateBranch.php (hunk 4 above)
Comment (leomarlo) [reply Lakat/src/SpecialCreateBranch.php:1]: Again, thanks a lot for your explanation.
